### PR TITLE
Core/Player: improve Quest & Phase updates

### PR DESCRIPTION
**Pull request by Palabola**
**Changes proposed:** Better functionality for Phase changing and updating.
-  Better handling for Quest based Phase changes. (Complete,Taken,Rewarded,(Objective Completed)...)

**Target branch(es):** 3.3.5/master

**Issues addressed:** Closes #

**Tests performed:** (Does it build, tested in-game, etc.)

**Known issues and TODO list:** None!

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -15792,6 +15792,7 @@ void Player::SendQuestUpdate(uint32 questId)
 
     UpdateForQuestWorldObjects();
     SendUpdatePhasing();
+	UpdateAreaAndZonePhase();
```
Comment (Keader) [inline src/server/game/Entities/Player/Player.cpp]: identation :)

**Review by Keader — CHANGES_REQUESTED** [review]
remove white space

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (Aokromes) [inline src/server/game/Entities/Player/Player.cpp]: in fact, tab to spaces :P

**Comment by jackpoz** [comment]
> Tests performed: (Does it build, tested in-game, etc.)

no tests performed ?

**Comment by zamalaev** [comment]
jackpoz, I checked on the Pandaren and works well

**Comment by DDuarte** [comment]
This is equal to #16973 which got open and closed by Lopfest.
@Lopfest could you give some info on this?

**Comment by funjoker** [comment]
U think blizz is doing those Updates via spell.
Saw many spells in sniffs which Force a Phase update

**Comment by jackpoz** [comment]
> jackpoz, I checked on the Pandaren and works well

you should write in the PR description which tests you did in the most detailed way. You should describe the previous behavior, the current behavior (and for this particular PR the quest ID you tested).

**Comment by Palabola** [comment]
@jackpoz  Easy,

**Current version:**
1. Create Phase in (phase_area) table and any Quest Condition(Reward,Complete,Taken...) for it into the (conditions) table.
2. Complete,Take or Reward the Quest and nothing will happen with your Phase. (It will update only  with your next relog/area change!)
**Updated version:**
1. Create Phase in (phase_area) table and any Quest Condition(Reward,Complete,Taken...) for it into the (conditions) table.
2. Complete,Take or Reward the Quest your Phase will be updated based on Phase_area and Condition.

**Proof of concept:**
- Long time ago same functionality were in spell_area table with quest_start_status / quest_end_status field. 
- 

**Posible conflicts:**
- None, since relog or re-enter the Area use the same effect.

**Comment by Lopfest** [comment]
@DDuarte Well, like I have written in that closed PR of mine, first I thought the same as this PR description sais. I dont remember exactly what changed my mind, but it looks like I found a quest that was not matching this behavior. I shoud have written the quest ID in the comment :(

**Comment by zamalaev** [comment]
I do not quite understand, but I think this hack

**Comment by Lopfest** [comment]
One great example of exception is this quest [The Missing Driver](http://www.wowhead.com/quest=29419/the-missing-driver) in pandaren starting zone. Phases should definitely not be updated automatically when the quest becomes complete. It should be updated manually via AI few moments after quest is complete, or MAYBE when player do a relog before the manual update.

**Comment by Palabola** [comment]
It's only update the phase if you allow in the condition to be updated, but okay If reloging after every quest more blizzlike up to you.

**Comment by Lopfest** [comment]
no forget relogging :D in that quest, you have condition that phase should update only if the quest is complete or rewarded. but the phase shouldnt be updated immediatelly after quest becomes complete, thats the point.

**Comment by funjoker** [comment]
@Lopfest 

> you have condition that phase should update only if the quest is complete or rewarded. but the phase shouldnt be updated immediatelly after quest becomes complete, thats the point.

example?

**Comment by Lopfest** [comment]
@funjoker 

> One great example of exception is this quest The Missing Driver in pandaren starting zone. Phases should definitely not be updated automatically when the quest becomes complete. It should be updated manually via AI few moments after quest is complete...

few posts above...

**Comment by Palabola** [comment]
@Lopfest Simlply not enought some quest ender become only seeable only after you completed the Quest just like some Phase which activeted after an Objective become completed. 
This PR work in practice, that's my last word about it.

**Comment by Lopfest** [comment]
can anyone make some statement about this PR and whether there can be added quest exception list in the code? For example something like this:
```
void Player::CompleteQuest(uint32 quest_id)
{
    if (quest_id)
    {
        switch (quest_id)
        {
            case 29419:
                SetQuestStatus(quest_id, QUEST_STATUS_COMPLETE, false);
                break;
            default:
                SetQuestStatus(quest_id, QUEST_STATUS_COMPLETE);
                break;
        }
...
```

**Comment by tkrokli** [comment]
@Lopfest : I am fairly sure it's a bad idea to hardcode the status of specific quests when you can use the DB table `disables` instead. Quest status should not be handled like spell exceptions.

**Comment by Lopfest** [comment]
I totaly forgot about disables table. That will suffice too.

**Comment by Aokromes** [comment]
Sooo, this pr is valid or invalid? it's ridiculous to have one 1 row PR opened for more than half year.

**Comment by funjoker** [comment]
for me it's invalid. Due to some quests do not force a phase update.
Phase updates are also often done by spells which are casted on quest accept.

**Comment by Shauren** [comment]
@funjoker going to trust you on this one, you have more experience with quest phasing

**Comment by funjoker** [comment]
The one and only thing i can imagine where the phase update could accure by quest update is, when a quest gets abonded.

**Comment by Golrag** [comment]
There is a quest flag in WPP UpdatePhaseShift (0x00400000). It's not in the core tho

**Pull request closed without merging**